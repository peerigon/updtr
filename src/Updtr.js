import EventEmitter from "events";
import path from "path";
import fs from "./util/fs";
import {
    USE_OPTIONS,
    UPDATE_TO_OPTIONS,
    SAVE_OPTIONS,
} from "./constants/config";
import exec from "./exec/exec";
import cmds from "./exec/cmds";
import parse from "./exec/parse";
import {
    RequiredOptionMissingError,
    OptionValueNotSupportedError,
    YarnWithCustomRegistryError,
} from "./errors";

// node v4 has no dedicated constants object.
// Remove this if node v4 is not supported anymore.
const FS_CONSTANTS = fs.constants === undefined ? fs : fs.constants;

function checkCwd(cwd) {
    if (typeof cwd !== "string") {
        throw new RequiredOptionMissingError("cwd", cwd);
    }
}

function checkUse(use) {
    if (USE_OPTIONS.indexOf(use) === -1) {
        throw new OptionValueNotSupportedError("use", use);
    }
}

function checkUpdateTo(updateTo) {
    if (UPDATE_TO_OPTIONS.indexOf(updateTo) === -1) {
        throw new OptionValueNotSupportedError("updateTo", updateTo);
    }
}

function checkSave(save) {
    if (SAVE_OPTIONS.indexOf(save) === -1) {
        throw new OptionValueNotSupportedError("save", save);
    }
}

function checkForYarnWithCustomReg(packageManager, registry) {
    if (packageManager === "yarn" && registry !== undefined) {
        throw new YarnWithCustomRegistryError();
    }
}

export default class Updtr extends EventEmitter {
    /**
     * The config passed-in here should look identically to the CLI config.
     * Dash-cased properties should be renamed to camelCased.
     * The goal is to replicate the API of the CLI as close as possible so users don't
     * have to guess the options.
     *
     * @param {UpdtrConfig} config
     */
    constructor(config) {
        super();

        const cwd = config.cwd;
        const registry = config.registry;
        const packageManager = config.use === undefined ?
            USE_OPTIONS[0] :
            config.use;
        const updateTo = config.updateTo === undefined ?
            UPDATE_TO_OPTIONS[0] :
            config.updateTo;
        const exclude = Array.isArray(config.exclude) === true ?
            config.exclude :
            [];
        const save = config.save === undefined ? SAVE_OPTIONS[0] : config.save;

        checkCwd(cwd);
        checkUse(packageManager);
        checkForYarnWithCustomReg(packageManager, registry);
        checkUpdateTo(updateTo);
        checkSave(save);

        this.config = {
            cwd,
            use: packageManager,
            exclude,
            test: config.test,
            registry,
            updateTo,
            save,
        };
        this.cmds = cmds[packageManager];
        this.parse = parse[packageManager];

        if (typeof config.test === "string") {
            this.cmds = {
                ...this.cmds,
                test: () => config.test,
            };
        }
    }
    async canAccessPackageJson() {
        let result = true;

        try {
            await fs.access(
                path.join(this.config.cwd, "package.json"),
                FS_CONSTANTS.R_OK | FS_CONSTANTS.W_OK // eslint-disable-line no-bitwise
            );
        } catch (err) {
            result = false;
        }

        return result;
    }
    exec(cmd) {
        return exec(this.config.cwd, cmd);
    }
    readFile(filenameInCwd) {
        return fs.readFile(path.join(this.config.cwd, filenameInCwd), "utf8");
    }
    writeFile(filenameInCwd, contents) {
        return fs.writeFile(
            path.join(this.config.cwd, filenameInCwd),
            contents
        );
    }
    dispose() {
        this.removeAllListeners();
    }
}
