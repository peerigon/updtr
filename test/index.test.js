import Updtr from "../src/Updtr";
import FakeUpdtr from "./helpers/FakeUpdtr";

const index = require("../src");

describe("index", () => {
    describe(".create", () => {
        it("should be a function", () => {
            expect(typeof index.create).toBe("function");
        });
        it("should return an instance of Updtr", () => {
            expect(index.create(FakeUpdtr.baseConfig)).toBeInstanceOf(Updtr);
        });
    });
    describe(".run", () => {
        it("should be a function", () => {
            expect(typeof index.run).toBe("function");
        });
    });
    describe(".errors", () => {
        it("should be an object with expected shape", () => {
            expect(index.errors).toMatchSnapshot();
        });
    });
});
