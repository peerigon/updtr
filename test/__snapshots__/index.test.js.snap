// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index .errors should be an object with expected shape 1`] = `
Object {
  "OptionValueNotSupportedError": [Function],
  "PackageJsonNoAccessError": [Function],
  "RequiredOptionMissingError": [Function],
  "YarnWithCustomRegistryError": [Function],
}
`;
